docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,171 @@
+```
+
+(the structure of the SDK and actual arguments is discussed below)
+
+However today, users are forced to run a sequence of commands tailored to only work with the proper directory structure on the system.
+
+## Major Goals
+
+1. Modularize the InstructLab workflow such that any part can be run independently
+2. Allow users to choose whether or not to take advantage of the config/system-profile method of running InstructLab. Meaning they do not need any pre-existing configuration to run the SDK.
+3. Standardize user contracts for the existing functionality of the InstructLab workflow. Existing CLI commands should be using the SDK once past click parsing, not separate code.
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
+
+1. Exposing all library functionality immediately
+2. Replacing CLI
+3. Shipping an SDK that is generally available as opposed to v1alpha1 or v1beta1.
+
+## Design
+
+### Versioning
+
+The SDK would start at version v1alpha1 such that it can change/break at any time for the first few iterations as libraries adjust their API surface.
+
+### Structure
+
+This SDK should live in a net new package inside of `instructlab/instructlab` preferably to limit unnecessary imports in a new repository. The SDK could be imported as `instructlab.core...`
+
+The user surface initially should look like this:
+
+`instructlab.core` contains all SDK definitions. Users can `from instructlab.core import...` to use specific SDK classes
+
+For most of the existing InstructLab command groups, there should be a class:
+
+`from instructlab.core import, InstructLab, Eval, Data, Model`
+
+However, things like configuration initialization, taxonomy management, system info are utilities. So they can live in a centralized `InstructLab` client class.
+
+```python
+
+ilab_client = InstructLab(config_path="", taxonomy="", auto_detect=True)
+
+diff = ilab_client.taxonomy.diff()
+
+if diff:
+    # not in v1alpha1
+    data_ingestor = Data.Ingestor(data_path="",...)
+    data_path = data_ingestor.ingest()
+    # not in v1alpha1
+
+    sdg_client = Data.Generator(client=openai_compat_client, data=data_path, teacher_model=path_to_model)
+
+    sdg_client.generate_data()
+
+    # you can either use a config obj or pass trainer args
+    trainer = Model.Trainer(student_model=path_to_student_model, configuration=configuration_object, training_args=...)
+
+    model_path = trainer.train_model()
+
+    # you can either use a config obj or each arg manually
+    eval_client = Eval.Evaluator(model_path=model_path, configuration=configuration_object...)
+
+    eval_client.mt_bench()
+```
+
+So the full list:
+
+```console
+instructlab.core.InstructLab
+instructlab.core.InstructLab.taxonomy.diff
+instructlab.core.InstructLab.system.info
+instructlab.core.InstructLab.config.show
+instructlab.core.InstructLab.config.get
```

Comment:
To illustrate my point above, I would prefer 
```suggestion
instructlab.core.Taxonomy.diff
instructlab.core.System.info
instructlab.core.Config.show
instructlab.core.Config.get
```

Reply:
This could be accomplished w/ `__init__.py` module visibility too.

Reply:
just as a note, these classes are departing from the structure of the other ones where its `Model.Trainer.train_model` for example, is this something we are ok with?

Reply:
Is this a departure from that? I don't see how

Reply:
this is outdated, the new structure aligns with this comment